app/controllers/shopify_app/callback_controller.rb

```diff
@@ -43,6 +49,25 @@ def respond_with_error
       redirect_to(login_url_with_optional_shop)
     end
 
+    def respond_with_user_token_flow
+      redirect_to(login_url_with_optional_shop)
+    end
+
+    def start_user_token_flow?(shopify_session)
+      return false unless ShopifyApp::SessionRepository.user_storage.present?
+      return false if shopify_session.online?
```

Comment:
how does `shopify_session.online?` check whether the response if for an online token?

Reply:
I think this is part of *Shopify API*, maybe this should be `is_online` instead? https://github.com/Shopify/shopify_api/blob/5a4785a40be012f7f3ec93ecb037508efff0c292/lib/shopify_api/auth/session.rb#L47

Reply:
Yes, I believe that is the original behaviour here - if we want online tokens, we go through OAuth twice in a row to get both of them.